apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -182,22 +196,85 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
     }
   }
 
-  public static loadFromFile(shrinkwrapYamlFilename: string): PnpmShrinkwrapFile | undefined {
+  public static loadFromFile(
+    shrinkwrapYamlFilename: string,
+    pnpmOptions: PnpmOptionsConfiguration
+  ): PnpmShrinkwrapFile | undefined {
     try {
       if (!FileSystem.exists(shrinkwrapYamlFilename)) {
         return undefined; // file does not exist
       }
 
       // We don't use JsonFile/jju here because shrinkwrap.json is a special NPM file format
       // and typically very large, so we want to load it the same way that NPM does.
-      const parsedData: IPnpmShrinkwrapYaml = yaml.safeLoad(FileSystem.readFile(shrinkwrapYamlFilename).toString());
+      const shrinkwrapContent: string = FileSystem.readFile(shrinkwrapYamlFilename).toString();
+      const parsedData: IPnpmShrinkwrapYaml = yaml.safeLoad(shrinkwrapContent);
+
+      let shrinkwrapHash: string | undefined;
+      if (pnpmOptions.preventManualShrinkwrapChanges) {
+        // Grab the shrinkwrap hash out of the comment where we store it.
+        const hashYamlCommentRegExp: RegExp = new RegExp(
+          `\\n\\s*${PnpmShrinkwrapFile._shrinkwrapHashPrefix}\\s*(\\S+)\\s*$`
+        );
+        const match: RegExpMatchArray | null = shrinkwrapContent.match(hashYamlCommentRegExp);
+        shrinkwrapHash = match ? match[1] : undefined;
+      }
 
-      return new PnpmShrinkwrapFile(parsedData, shrinkwrapYamlFilename);
+      return new PnpmShrinkwrapFile(
+        parsedData,
+        shrinkwrapYamlFilename,
+        shrinkwrapHash,
+        pnpmOptions.preventManualShrinkwrapChanges
+      );
     } catch (error) {
       throw new Error(`Error reading "${shrinkwrapYamlFilename}":${os.EOL}  ${error.message}`);
     }
   }
 
+  /** @override */
+  public shouldForceRecheck(): boolean {
+    // Ensure the shrinkwrap is rechecked when the hash is enabled but no hash is populated.
+    return super.shouldForceRecheck() || (!!this._shrinkwrapHashEnabled && !this._shrinkwrapHash);
+  }
+
+  /** @override */
+  public validate(
+    packageManagerOptionsConfig: PackageManagerOptionsConfigurationBase,
+    policyOptions: IPolicyValidatorOptions
+  ) : void {
+    super.validate(packageManagerOptionsConfig, policyOptions);
+    if (!(packageManagerOptionsConfig instanceof PnpmOptionsConfiguration)) {
```

Comment:
Would this ever be the case? Should we throw in this case?

Reply:
Throwing probably makes the most sense. We should never be given the incorrect options if the shrinkwrap cannot validate against them